evm/src/executor/inspector/cheatcodes/expect.rs

```diff
@@ -123,46 +123,79 @@ pub struct ExpectedEmit {
 }
 
 pub fn handle_expect_emit(state: &mut Cheatcodes, log: RawLog, address: &Address) {
-    // Fill or check the expected emits
-    if let Some(next_expect_to_fill) =
-        state.expected_emits.iter_mut().find(|expect| expect.log.is_none())
-    {
-        // We have unfilled expects, so we fill the first one
-        next_expect_to_fill.log = Some(log);
-    } else if let Some(next_expect) = state.expected_emits.iter_mut().find(|expect| !expect.found) {
-        // We do not have unfilled expects, so we try to match this log with the first unfound
-        // log that we expect
-        let expected =
-            next_expect.log.as_ref().expect("we should have a log to compare against here");
-
-        let expected_topic_0 = expected.topics.get(0);
-        let log_topic_0 = log.topics.get(0);
-
-        // same topic0 and equal number of topics should be verified further, others are a no
-        // match
-        if expected_topic_0
-            .zip(log_topic_0)
-            .map_or(false, |(a, b)| a == b && expected.topics.len() == log.topics.len())
-        {
-            // Match topics
-            next_expect.found = log
-                .topics
-                .iter()
-                .skip(1)
-                .enumerate()
-                .filter(|(i, _)| next_expect.checks[*i])
-                .all(|(i, topic)| topic == &expected.topics[i + 1]);
-
-            // Maybe match source address
-            if let Some(addr) = next_expect.address {
-                next_expect.found &= addr == *address;
+    // Fill or check the expected emits.
+    // We expect for emit checks to be filled as they're declared (from oldest to newest),
+    // so we fill them and push them to the back of the queue.
+    // If the user has properly filled all the emits, they'll end up in their original order.
+    // If not, the queue will not be in the order the events will be intended to be filled,
+    // and we'll be able to later detect this and bail.
+
+    // First, we can return early if all events have been matched.
+    // This allows a contract to arbitrarily emit more events than expected (additive behavior),
+    // as long as all the previous events were matched in the order they were expected to be.
+    if !state.expected_emits.iter().any(|expected| !expected.found) {
```

Comment:
this is the same as `if state.expected_emits.iter().all(|expected| expected.found)`, right?

Reply:
yep indeed

Reply:
done in https://github.com/foundry-rs/foundry/pull/4920/commits/d72a935f944a7a823e9c448dbdf39c6cbeb9151f , waiting for ci